src/dataflow/src/source/kafka.rs

```diff
@@ -121,44 +116,14 @@ impl SourceInfo<Vec<u8>> for KafkaSourceInfo {
             .unwrap()
             .offset;
 
-        // For transactional or compacted topics, the "last_offset" may not correspond to
-        // the last record in the topic (either because it has been GCed or because
-        // it corresponds to an abort/commit marker).
-        // topic and partition entries are not guaranteed to exist if the poll request to metadata has not succeeded:
-        // In Kafka, the position of the consumer is set to the offset *after* the last offset that the consumer has
-        // processed, so we have to decrement it by one to get the last processed offset
-
-        // We separate these two cases, as consumer.position() is an expensive call that should
-        // be avoided if possible. Case 1 and 2.a occur first, and we only test 2.b when necessary
-        if !self.has_partition(kafka_pid) // Case 1
-        || last_offset >= offset
-        // Case 2.a
-        {
+        // For transactional or compacted topics, the `last_offset` may not correspond to
+        // an accurate representation of how far we have read up to (because subsequent offsets
+        // have been garbage collected). However, `last_offset` still represents the lower bound
+        // for the value of the next offset in this partition.
+        if !self.has_partition(kafka_pid) || last_offset >= offset {
             true
         } else {
-            let mut current_consumer_position: MzOffset = KafkaOffset {
```

Comment:
I'm not 100% sure about this block of changes otherwise I feel good about everything else in this pr

Reply:
my understanding is that the previous code to check the current kafka read position was a performance optimization and not required for correctness but I will try to validate that more carefully

Reply:
the current logic was introduced in https://github.com/MaterializeInc/materialize/pull/3259

Reply:
@elindsey let me know if you have thoughts here

Reply:
Ok I did some slack sleuthing I think the original motivation for the "check current position" logic came from https://github.com/MaterializeInc/materialize/issues/3265 but I'm not sure if thats something to be concerned about anymore

Reply:
I'm pretty sure none of this is applicable with rt. rt in its current form doesn't care about offsets at all